File: python/structural/11_flyweight.py
"""
设计模式: 享元模式 (Flyweight)
分类: 结构型

意图: 运用共享技术高效支持大量细粒度对象，减少内存占用。
核心思想: 分离内部状态（共享、不变）与外部状态（不共享、调用者传入）。
编程范式: OOP — 对象池/缓存 + 内外状态分离。
"""

from __future__ import annotations
from dataclasses import dataclass


# ── 享元：内部状态（共享）────────────────────────────────────────────────
@dataclass(frozen=True)
class TreeType:
    name: str
    color: str
    texture: str

    def draw(self, x: int, y: int) -> None:
        print(f"Tree[{self.name}/{self.color}] at ({x},{y})")


# ── 享元工厂 ──────────────────────────────────────────────────────────────
class TreeFactory:
    _types: dict[str, TreeType] = {}

    @classmethod
    def get(cls, name: str, color: str, texture: str) -> TreeType:
        key = f"{name}_{color}"
        if key not in cls._types:
            cls._types[key] = TreeType(name, color, texture)
            print(f"[Factory] created new TreeType: {key}")
        return cls._types[key]

    @classmethod
    def type_count(cls) -> int:
        return len(cls._types)


# ── 树实例：携带外部状态 ──────────────────────────────────────────────────
@dataclass
class Tree:
    x: int
    y: int
    tree_type: TreeType

    def draw(self) -> None:
        self.tree_type.draw(self.x, self.y)


# ── 演示 ──────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    forest: list[Tree] = []
    for i in range(5):
        forest.append(Tree(i*10, i*5, TreeFactory.get("oak",   "green", "rough")))
        forest.append(Tree(i*7,  i*3, TreeFactory.get("pine",  "dark",  "smooth")))
        forest.append(Tree(i*3,  i*9, TreeFactory.get("birch", "white", "thin")))

    for tree in forest:
        tree.draw()

    print(f"TreeType instances: {TreeFactory.type_count()} (vs {len(forest)} trees)")

File: cpp/structural/11_flyweight.cpp
/**
 * 设计模式: 享元模式 (Flyweight)
 * 分类: 结构型
 *
 * 意图: 运用共享技术高效支持大量细粒度对象，减少内存占用。
 * 核心思想: 分离内部状态（共享，不变）与外部状态（不共享，由调用者传入）。
 * 编程范式: OOP — 对象池/缓存 + 内外状态分离。
 */

#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>

// ── 享元：内部状态（共享）────────────────────────────────────────────────
struct TreeType {
    std::string name;
    std::string color;
    std::string texture;

    void draw(int x, int y) const {
        std::cout << "Tree[" << name << "/" << color << "] at (" << x << "," << y << ")\n";
    }
};

// ── 享元工厂 ──────────────────────────────────────────────────────────────
class TreeFactory {
    std::unordered_map<std::string, std::shared_ptr<TreeType>> types_;
public:
    std::shared_ptr<TreeType> get(const std::string& name,
                                   const std::string& color,
                                   const std::string& texture) {
        std::string key = name + "_" + color;
        if (!types_.count(key)) {
            types_[key] = std::make_shared<TreeType>(TreeType{name, color, texture});
            std::cout << "[Factory] created new TreeType: " << key << "\n";
        }
        return types_[key];
    }
    size_t typeCount() const { return types_.size(); }
};

// ── 树实例：携带外部状态（x, y 坐标）────────────────────────────────────
struct Tree {
    int x, y;
    std::shared_ptr<TreeType> type;
    void draw() const { type->draw(x, y); }
};

int main() {
    TreeFactory factory;
    std::vector<Tree> forest;

    // 种 1000 棵树，但只有 3 种 TreeType
    for (int i = 0; i < 5; ++i) {
        forest.push_back({i * 10, i * 5, factory.get("oak",   "green",  "rough")});
        forest.push_back({i * 7,  i * 3, factory.get("pine",  "dark",   "smooth")});
        forest.push_back({i * 3,  i * 9, factory.get("birch", "white",  "thin")});
    }

    for (auto& tree : forest) tree.draw();
    std::cout << "TreeType instances: " << factory.typeCount() << " (vs " << forest.size() << " trees)\n";
    return 0;
}
